File: packages/domain/src/reward.ts
import { fail } from "./errors.ts";
import { addU64, asU64 } from "./math.ts";

export interface RewardSplit {
  /** Reward for every epoch except the last. */
  readonly base: bigint;
  /** Remainder added to the last epoch only, so the maximum earnable equals the accepted bid exactly. */
  readonly extra: bigint;
}

/** `base = floor(R / N)`, `extra = R mod N`. */
export function splitReward(acceptedRewardRaw: bigint, totalEpochs: number): RewardSplit {
  if (totalEpochs === 0) return fail("DivisionByZero", "zero epochs");
  const n = BigInt(totalEpochs);
  const reward = asU64(acceptedRewardRaw);
  return { base: reward / n, extra: reward % n };
}

/** Reward a compliant epoch earns. Epochs `0..N-2` earn `base`; the final epoch earns `base + extra`. */
export function epochReward(acceptedRewardRaw: bigint, totalEpochs: number, index: number): bigint {
  const { base, extra } = splitReward(acceptedRewardRaw, totalEpochs);
  if (!Number.isSafeInteger(index) || index < 0 || index >= totalEpochs)
    return fail("EpochOutOfRange");
  return index === totalEpochs - 1 ? addU64(base, extra) : base;
}

File: crates/mandate-core/src/reward.rs
//! Exact reward split. `base = floor(R / N)`, `extra = R mod N`; epochs `0..N-2` earn `base` and the
//! final epoch earns `base + extra`, so the maximum earnable equals the accepted reward exactly.
use crate::math::{add_u64, div_u64, rem_u64};
use crate::{MandateCoreError, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RewardSplit {
    pub base: u64,
    pub extra: u64,
}

pub fn split_reward(accepted_reward_raw: u64, total_epochs: u32) -> Result<RewardSplit> {
    let n = u64::from(total_epochs);
    Ok(RewardSplit {
        base: div_u64(accepted_reward_raw, n)?,
        extra: rem_u64(accepted_reward_raw, n)?,
    })
}

/// Reward a compliant epoch earns.
pub fn epoch_reward(accepted_reward_raw: u64, total_epochs: u32, index: u32) -> Result<u64> {
    let split = split_reward(accepted_reward_raw, total_epochs)?;
    if index >= total_epochs {
        return Err(MandateCoreError::EpochOutOfRange);
    }
    // `index < total_epochs`, so `total_epochs >= 1` and this subtraction cannot underflow.
    let last = total_epochs
        .checked_sub(1)
        .ok_or(MandateCoreError::EpochOutOfRange)?;
    if index == last {
        add_u64(split.base, split.extra)
    } else {
        Ok(split.base)
    }
}
